### ISSUE TITLE
Set up shrinkpack to see how it affects CI times

### ISSUE THREAD

**Issue by jmeas** [jmeas/moolah#369]
[shrinkpack](https://github.com/JamieMason/shrinkpack) could help speed up the ~6min build times I'm seeing

**Comment by jmeas** [comment]
This exploded Travis. I'll have to check back and try again another time.

**Comment by jmeas** [comment]
I wasn't setting the `--dev` for shrinkwrap, so it was only getting `deps` and not `devDeps`.

**Comment by jmeas** [comment]
Shrinkpack is a pain to get working, so I'm gonna close as wontfix for now.

**Comment by JamieMason** [comment]
Please raise issues over at https://github.com/JamieMason/shrinkpack/issues @jmeas and I'll do my best to try and help you 👍

**Comment by JamieMason** [comment]
I'll clone moolah now actually and see if I can get it working.

**Comment by JamieMason** [comment]
damn, `shrinkpack` failed for me because I can't resolve the packages under https://lunabuild.akamai.com/nexus – looks to be a private repository. 

![private-nexus](https://cloud.githubusercontent.com/assets/320492/16837998/c2f07e56-49be-11e6-83d4-476bff64ad02.png)

Presumably you can hit that though from your network.

One thing I did notice when I deleted npm-shrinkwrap.json and ran `npm shrinkwrap --dev` again, there were a lot of packages missing ([diff.txt](https://github.com/jmeas/moolah/files/363582/diff.txt) attached). That could have been part of the problem?

**Comment by jmeas** [comment]
@JamieMason , thanks for the offer to help! I'll be able to respond better later tonight, but for now I'll reopen this issue.

> damn, shrinkpack failed for me because I can't resolve the packages under https://lunabuild.akamai.com/nexus – looks to be a private repository.

That's really odd – I have no idea what that could be. I'm not able to hit that URL from my network. This project has no private deps that I know of, and Travis CI (the free version) has had no problem installing the deps in the package.json to run the tests. Also, Heroku has had no problem building and running the app.

> One thing I did notice when I deleted npm-shrinkwrap.json and ran npm shrinkwrap --dev again, there were a lot of packages missing (diff.txt attached). That could have been part of the problem?

Interesting. I'll take a closer look at my dependencies and see if I can make sense of the discrepancy.

---

Thanks again! :v:

**Comment by JamieMason** [comment]
I see npm is version 3.10.3, there was a regression in npm which was fixed
in 3.10.4, can you upgrade and retry?

**Comment by jmeas** [comment]
@JamieMason I'm trying that now. I also realized that I had my `nvm` set to `6.2.2` instead of `6.3.0` when I made the other shrinkpack, which may have affected things.

All the versions are lined up now, and I think I made it so Travis will install the latest npm. Fingers crossed!

**Comment by jmeas** [comment]
That seemed to have fixed the problem I was seeing before. Travis failed on linting since I need to rebase off of master, and Heroku exploded because it's using an old npm.

:v:

**Comment by JamieMason** [comment]
> I also realized that I had my nvm set to 6.2.2 instead of 6.3.0 when I made the other shrinkpack, which may have affected things

I _think_ that would've been fine, `node_shrinkwrap/*` would've contained the same tarballs either way.

> That seemed to have fixed the problem I was seeing before. 

Boom! 💥  any problems in the future please let me know, real world usage like this helps me spot and issues out in the wild 👍  Thanks for using shrinkpack.

**Comment by jmeas** [comment]
Sure thing :)

Thanks for building such a useful tool!

**Comment by jmeas** [comment]
Gonna reopen because heroku feature branch deploys have been broken since this landed. master branch deploys seem to work though?

**Comment by JamieMason** [comment]
No worries @jmeas, if you need any input from me please add more info here so I can try and help out.

**Comment by jmeas** [comment]
Thanks for the offer to help, @JamieMason ! Sorry it's taken a bit – I've been busy with things.

The tl;dr is that there's a thing called a build pipeline in Heroku. PRs that I create automatically create "review apps," which are deployed versions of the app. Also, pushes to master deploy to the test environment. Pushes to master work, but review apps don't work.

<details>
<summary>View truncated build fail log (click to expand)</summary>
```sh
-----> Deleting 2 files matching .slugignore patterns.
-----> Node.js app detected
-----> Creating runtime environment
       
       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=false
       NODE_ENV=production
       NODE_MODULES_CACHE=true
-----> Installing binaries
       engines.node (package.json):  6.3.0
       engines.npm (package.json):   unspecified (use default)
       
       Downloading and installing node 6.3.0...
       Using default npm version: 3.10.3
-----> Restoring cache
       Skipping cache restore (new runtime signature)
-----> Building dependencies
       Installing node modules (package.json + shrinkwrap)
       npm ERR! addLocal Could not install /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/dom-serializer/node_shrinkwrap/domelementtype-1.1.3.tgz
       npm ERR! addLocal Could not install /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/esrecurse/node_shrinkwrap/estraverse-4.1.1.tgz
       npm ERR! addLocal Could not install /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/merge-stream/node_shrinkwrap/readable-stream-2.1.4.tgz
       npm ERR! addLocal Could not install /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/got/node_shrinkwrap/object-assign-3.0.0.tgz
       npm ERR! addLocal Could not install /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/stylelint/node_shrinkwrap/resolve-from-2.0.0.tgz
       npm ERR! Linux 3.13.0-91-generic
       npm ERR! argv "/tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/.heroku/node/bin/node" "/tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/.heroku/node/bin/npm" "install" "--unsafe-perm" "--userconfig" "/tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/.npmrc"
       npm ERR! node v6.3.0
       npm ERR! npm  v3.10.3
       npm ERR! path /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/dom-serializer/node_shrinkwrap/domelementtype-1.1.3.tgz
       npm ERR! code ENOENT
       npm ERR! errno -2
       npm ERR! syscall open
       
       npm ERR! enoent ENOENT: no such file or directory, open '/tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/dom-serializer/node_shrinkwrap/domelementtype-1.1.3.tgz'
       npm ERR! enoent ENOENT: no such file or directory, open '/tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/node_modules/dom-serializer/node_shrinkwrap/domelementtype-1.1.3.tgz'
       npm ERR! enoent This is most likely not a problem with npm itself
       npm ERR! enoent and is related to npm not being able to find a file.
       npm ERR! enoent
       
       npm ERR! Please include the following file with any support request:
       npm ERR!     /tmp/build_925c1e682aefd67272ff87bee7714a64/jmeas-moolah-f7c2dde/npm-debug.log
-----> Build failed
```
</details>

(I didn't include all of the `addLocal` errors; only some at the top and bottom of the log).

Yet when deploying to master, it works.

---

I thought this might be due to dep bumps in package.json messing with shrinkwrap, but it doesn't seem to be the case as all PRs ([like this one](https://github.com/jmeas/moolah/pull/393)) fail.

This is the weirdest thing. I don't think there's any difference between the review app env and the test environment, yet one consistently fails and the other consistently passes. This might be an issue with some difference I don't know about, so I'll keep looking around.

Don't spend too much time on this if you think this is a Heroku specific bug rather than a shrinkpack specific bug ;)

**Comment by jmeas** [comment]
<details>
  <summary>Successful log here</summary>

```sh
-----> Deleting 2 files matching .slugignore patterns.
-----> Node.js app detected
-----> Creating runtime environment
       
       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=false
       NODE_ENV=production
       NODE_MODULES_CACHE=true
-----> Installing binaries
       engines.node (package.json):  6.3.0
       engines.npm (package.json):   unspecified (use default)
       
       Downloading and installing node 6.3.0...
       Using default npm version: 3.10.3
-----> Restoring cache
       Loading 2 from cacheDirectories (default):
       - node_modules
       - bower_components (not cached - skipping)
-----> Building dependencies
       Installing node modules (package.json + shrinkwrap)
       
       > moolah@0.0.1 postinstall /tmp/build_a7de7759e07ebe79b43624c85a63117d/jmeas-moolah-b4434db
       > if [ $BUILD_ASSETS ]; then npm run build-assets; fi
       
       
       > moolah@0.0.1 build-assets /tmp/build_a7de7759e07ebe79b43624c85a63117d/jmeas-moolah-b4434db
       > gulp build
       
```
</details>

here's a successful build log. Still reviewing the differences myself :v:

**Comment by jmeas** [comment]
Gonna close as revisit 👌

**Comment by JamieMason** [comment]
What surprises me is that it says you're using npm 3.10.3, which shouldn't
work l due to an npm regression fixed in 3.10.4. could you try upgrading
npm?

**Comment by jmeas** [comment]
Ahh, interesting. I didn't even catch that, even though you pointed it out before for Travis. I'll try that soon 👌

**Comment by jmeas** [comment]
Including shrinkpack seems to have caused: https://github.com/jmeas/moolah/issues/433

The cause is being tracked upstream at: https://github.com/JamieMason/shrinkpack/issues/40
